File: NeetCode/Backtracking/Permutations/permutations.py
from typing import Optional, List


class Solution:
    def permute(self, nums: List[int]) -> List[List[int]]:
        results = []

        if len(nums) == 0:
            return [[]]

        if len(nums) == 1:
            return [nums]

        for index in range(len(nums)):

            array = nums.copy()
            curr = array.pop(index)
            sub = self.permute(array)

            for s in sub:
                results.append([curr, *s])

        return results


if __name__ == "__main__":
    sol = Solution()
    nums = [1, 2, 3]

    print(sol.permute(nums))
    print("Running Solution...")

File: NeetCode/Backtracking/Permutations/Permutations.cpp
// Permutations.cpp
#include <algorithm>
#include <array>
#include <cassert>
#include <cctype>
#include <climits>
#include <cmath>
#include <deque>
#include <functional>
#include <iostream>
#include <map>
#include <numeric>
#include <queue>
#include <set>
#include <stack>
#include <string>
#include <unordered_map>
#include <unordered_set>
#include <utility>
#include <vector>

using namespace std;

class Solution
{
public:
    vector<vector<int>> permute(vector<int> &nums)
    {
        vector<vector<int>> results;

        if (nums.size() == 0)
        {
            return results;
        }

        if (nums.size() == 1)
        {
            results.push_back(nums);
            return results;
        }

        for (int index = 0; index < nums.size(); index++)
        {
            vector<int> arr = nums;
            int curr = arr[index];
            arr.erase(arr.begin() + index);
            vector<vector<int>> tmp = permute(arr);

            for (vector<int> sub : tmp)
            {
                sub.insert(sub.begin(), curr);
                results.push_back(sub);
            }
        }
        return results;
    }
};

int main()
{
    Solution sol;

    vector<int> n = {1, 2, 3};

    vector<vector<int>> r = sol.permute(n);

    for (vector<int> p : r)
    {
        for (int i : p)
        {
            cout << i << " ";
        }
        cout << "\n";
    }
}
